test/libsolidity/ASTRecoveryTest.cpp

```diff
@@ -0,0 +1,143 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <test/libsolidity/ASTRecoveryTest.h>
+#include <test/Options.h>
+#include <libdevcore/AnsiColorized.h>
+#include <liblangutil/SourceReferenceFormatterHuman.h>
+#include <libsolidity/ast/ASTJsonConverter.h>
+#include <libsolidity/interface/CompilerStack.h>
+#include <boost/algorithm/string.hpp>
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/throw_exception.hpp>
+#include <fstream>
+#include <memory>
+#include <stdexcept>
+
+using namespace langutil;
+using namespace dev::solidity;
+using namespace dev::solidity::test;
+using namespace dev::formatting;
+using namespace dev;
+using namespace std;
+namespace fs = boost::filesystem;
+using namespace boost::unit_test;
+
+ASTRecoveryTest::ASTRecoveryTest(string const& _filename)
```

Comment:
This class and the class `ASTJsonTest` are almost identical, only a few things missing in this one.
Do you think this will be modified? If not, I am thinking it makes sense that this one could become a base-class for `ASTJsonTest` to avoid code duplication.

Reply:
i think it likely it would be modified to handle specific patterns of error recovery that I don't doesn't occur in ASTJsonTest. 

Donald Knuth has is noted to have said: 

> premature optimization is the root of evil 

And Richard Feynman is noted to have said that you can unify any theory with any other theory by adding more parameters.

At this stage what is important is to get this out there, see how it works, and then adjust things. We really have no idea how things will need to be adjusted. So compacting, optimizing and streamlining at this point is just going to create more complexity if not additional work to compact and then uncompact, and parameterize when it needs modification.

As a very small example, note that ASTJsonTest also doesn't have a single comment in it as I just added in bd6d3c675.  So that comment would be then also need to be adjusted somehow to say that it tests this or that. 

If you want I can put in a `TODO` or a `FIXME`, or whatever the convention is for noting something to revisit later, I can do that now. But if so, let me know what the convention is.